net/socks5/socks5.go

```diff
@@ -199,19 +200,30 @@ func (c *Conn) handleRequest() error {
 	errc := make(chan error, 2)
 	go func() {
 		_, err := io.Copy(c.clientConn, srv)
-		if err != nil {
+		if errors.Is(err, io.EOF) {
+			err = nil
+		} else if err != nil {
 			err = fmt.Errorf("from backend to client: %w", err)
 		}
 		errc <- err
 	}()
 	go func() {
 		_, err := io.Copy(srv, c.clientConn)
-		if err != nil {
+		if errors.Is(err, io.EOF) {
+			err = nil
+		} else if err != nil {
 			err = fmt.Errorf("from client to backend: %w", err)
 		}
 		errc <- err
 	}()
-	return <-errc
+	err1 := <-errc
```

Comment:
The normal pattern for these stitch-both-ways-together copies is to return on first error, else you risk blocking forever.

The cleanup happens by deferred closes after/as the first error is returned.